smallclib/Makefile

```diff
@@ -0,0 +1,140 @@
+
+# Explicitly picking files to compile.
+LIB_SOURCES  = ctype/ctype_.c
+LIB_SOURCES += ctype/isalnum.c
+LIB_SOURCES += ctype/isalpha.c
+LIB_SOURCES += ctype/isascii.c
+LIB_SOURCES += ctype/isblank.c
+LIB_SOURCES += ctype/iscntrl.c
+LIB_SOURCES += ctype/isdigit.c
+LIB_SOURCES += ctype/isgraph.c
+LIB_SOURCES += ctype/islower.c
+LIB_SOURCES += ctype/isprint.c
+LIB_SOURCES += ctype/ispunct.c
+LIB_SOURCES += ctype/isspace.c
+LIB_SOURCES += ctype/isupper.c
+LIB_SOURCES += errno/errno.c
+LIB_SOURCES += math/s_fpclassify.c
+LIB_SOURCES += math/s_frexp.c
+LIB_SOURCES += misc/fini.c
+LIB_SOURCES += misc/init.c
+LIB_SOURCES += misc/setjmp.S
+LIB_SOURCES += stdio/___low_sprintf.c
+LIB_SOURCES += stdio/__format_parser.c
+LIB_SOURCES += stdio/snprintf.c
+LIB_SOURCES += stdio/sprintf.c
+LIB_SOURCES += stdlib/abs.c
+LIB_SOURCES += stdlib/atoi.c
+LIB_SOURCES += stdlib/atol.c
+LIB_SOURCES += stdlib/atoll.c
+LIB_SOURCES += stdlib/bsearch.c
+LIB_SOURCES += stdlib/qsort.c
+LIB_SOURCES += stdlib/rand.c
+LIB_SOURCES += stdlib/rand_data.c
+LIB_SOURCES += stdlib/strtod.c
+LIB_SOURCES += stdlib/strtol.c
+LIB_SOURCES += stdlib/strtoll.c
+LIB_SOURCES += string/bzero.c
+LIB_SOURCES += string/memchr.c
+LIB_SOURCES += string/memcpy.c
+LIB_SOURCES += string/memmove.c
+LIB_SOURCES += string/memset.c
+LIB_SOURCES += string/strcmp.c
+LIB_SOURCES += string/strcpy.c
+LIB_SOURCES += string/strerror.c
+LIB_SOURCES += string/strlen.c
+LIB_SOURCES += string/strncmp.c
+LIB_SOURCES += string/strspn.c
+LIB_SOURCES += string/strstr.c
+
+LIB_SOURCES_FULL = $(addprefix $(LIB_PATH)/, $(LIB_SOURCES))
+
+# LIB_SOURCES += stdlib/wctomb.c
+# NOTE: wctomb IS needed by __format_parser, but it requires the whole
+# locale as a dependency, and even environemnt access. Instead, we'll
+# provide a custom wctomb which fails to represent any wide character.
+
+# Extra utilities
+EXTRA_SOURCES += extra/wctomb.c
+EXTRA_SOURCES += extra/kperror.c
+EXTRA_SOURCES += extra/kputs.c
+EXTRA_SOURCES += extra/kprintf.c
+
+# The default rule.
+all: $(LIBNAME).a
+
+# This rule creates the build directory.
+$(LIB_BUILD_DIR):
+	mkdir -p $(LIB_BUILD_DIR)
+
+# This rule creates the library archive. To define it, we need a full
+# list of all object files.
+LIB_OBJECTS  = \
+	$(addprefix $(LIB_BUILD_DIR)/,\
+		$(patsubst %.S,%.o,             \
+		$(patsubst %.c,%.o,             \
+			$(notdir $(LIB_SOURCES))    \
+		) \
+		) \
+	)
+EXTRA_OBJECTS = \
+	$(patsubst %.S,%.o, \
+	$(patsubst %.c,%.o, \
+		$(EXTRA_SOURCES) \
+	) \
+	)
+$(LIBNAME).a: $(EXTRA_OBJECTS) $(LIB_OBJECTS)
+	@echo "Creating $@"
+	$(AR) rvs $@ $(EXTRA_OBJECTS) $(LIB_OBJECTS) > /dev/null
+
+# Generate rules for each library file separately. A simple pattern
+# rule will not suffice, because we need to correct the file directory
+# for rule and dependency.
+define emit_lib_rule =
+CFILE := $(1)
+OFILE := $(patsubst %.S,%.o,$(patsubst %.c,%.o,$(notdir $(1))))
+$(LIB_BUILD_DIR)/$(OFILE): $(LIB_PATH)/$(CFILE) | $(LIB_BUILD_DIR)
```

Comment:
Nice GNU Make magic :)

Please add some reference to the documentation:
* http://www.gnu.org/software/make/manual/make.html#Prerequisite-Types
* http://www.gnu.org/software/make/manual/make.html#Eval-Function